src/typing/type_inference_js.ml

```diff
@@ -5939,38 +6076,40 @@ and mk_body id cx type_params_map ~kind ?(derived_ctor=false)
   Env_js.havoc_all();
 
   (* create and prepopulate function scope *)
-  let function_scope =
-    let scope = Scope.fresh ~var_scope_kind:kind () in
-    (* add param bindings *)
-    param_types_map |> SMap.iter (fun name t ->
-      let loc = match SMap.get name param_locs_map with
-        | Some loc -> loc
-        | None -> loc_of_t t
-      in
-      let entry = Scope.Entry.new_var ~loc t in
-      Scope.add_entry name entry scope
-    );
-    (* early-add our own name binding for recursive calls *)
-    (match id with
-    | None -> ()
-    | Some (loc, { Ast.Identifier.name; _ }) ->
-      let entry = Scope.Entry.new_var ~loc (AnyT.at loc) in
-      Scope.add_entry name entry scope);
-    (* special bindings for this, super, and return value slot *)
-    initialize_this_super derived_ctor this super scope;
-    Scope.(
-      let new_entry t =
-        Entry.(new_const ~loc:(loc_of_t t) ~state:Initialized t)
-      in
-      add_entry (internal_name "yield") (new_entry yield) scope;
-      add_entry (internal_name "next") (new_entry next) scope;
-      add_entry (internal_name "return") (new_entry ret) scope
-    );
-    scope
-  in
+  let function_scope = Scope.fresh ~var_scope_kind:kind () in
 
+  (* push the scope early so default exprs can reference earlier params *)
   Env_js.push_var_scope cx function_scope;
 
+  (* add param bindings *)
+  params |> Params.iter Scope.(fun (name, t, loc) ->
+    (* add default value as lower bound, if provided *)
+    Params.with_default name (fun expr ->
+      let te = expression cx type_params_map expr in
+      Flow_js.flow cx (te, t)
+    ) params;
+    (* add to scope *)
+    let entry = Entry.new_var t ~loc in
```

Comment:
IIUC, we should use `let` bindings here to prevent argument name clash, even though doing so would not prevent the static error for function body `let`s (due to separate scopes)? I'm fine w/ that, just want to make sure we don't lose track of that static error either.